### PR TITLE
Refine reading state approach

### LINKED ISSUE #1765: Migrate to new approach to read program state

### PR DESCRIPTION
**Pull request by breathx**
Resolves #1765

### CODE REVIEW

File: core/src/code.rs [hunk 1]
```diff
@@ -219,36 +221,95 @@ impl Code {
...
+                exports,
+                static_pages,
+                instruction_weights_version: version,
+            })
         }
     }
 
+    /// Create the code with instrumentation, but without checks.
+    /// There is also no check for static memory pages.
+    pub fn new_raw_with_rules<R, GetRulesFn>(
```
Comment (ark0f) [inline core/src/code.rs:258]: Testing required

File: scripts/read-state.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
...00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
```
Comment (ark0f) [inline scripts/read-state.sh:16]: Move away the whole binary from script

**Review by ark0f — APPROVED** [review]
To the future:
- Uncomment and rewrite tests
- Add UI tests for macro
- Resolve TODOs

**Review by osipov-mit — APPROVED** [review]

File: pallets/gear/rpc/src/lib.rs [hunk 3]
```diff
@@ -303,4 +319,52 @@ where
...
+    fn read_metahash(
+        &self,
+        program_id: H256,
+        at: Option<<Block as BlockT>::Hash>,
+    ) -> RpcResult<H256> {
+        let at = BlockId::hash(at.unwrap_or_else(||
+            // If the block hash is not supplied assume the best block.
+            self.client.info().best_hash));
+
+        self.run_with_api_copy(|api| api.read_metahash(&at, program_id))
```
Comment (gshep) [inline pallets/gear/rpc/src/lib.rs:368]: I'm not sure if these methods need to be run within copied context.

**Review by gshep — APPROVED** [review]
Discussed this PR. It is ok to merge it in that state and then have a subsequent refactoring.

Good work! :+1:

File: core/src/message/mod.rs [hunk 4]
```diff
@@ -109,9 +109,32 @@ pub enum DispatchKind {
     Signal,
 }
 
-impl DispatchKind {
-    /// Convert DispatchKind into entry point function name.
-    pub fn into_entry(self) -> &'static str {
+/// Trait defining type could be used as entry point for a wasm module.
+pub trait WasmEntry: Sized {
```
[the remark this answers is not in the document]
Comment (breathx) [reply core/src/message/mod.rs:113]: Discussed

File: pallets/gear/rpc/runtime-api/src/lib.rs [hunk 5]
```diff
@@ -30,5 +30,11 @@ sp_api::decl_runtime_apis! {
 
         /// Generate inherent-like extrinsic the runs message queue processing.
         fn gear_run_extrinsic() -> <Block as BlockT>::Extrinsic;
+
+        fn read_state(program_id: H256) -> Result<Vec<u8>, Vec<u8>>;
+
+        fn read_state_using_wasm(program_id: H256, fn_name: Vec<u8>, wasm: Vec<u8>, argument: Option<Vec<u8>>) -> Result<Vec<u8>, Vec<u8>>;
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/rpc/runtime-api/src/lib.rs:36]: discussed

File: utils/wasm-builder/src/wasm_project.rs [hunk 6]
```diff
@@ -214,45 +247,91 @@ impl WasmProject {
         // Remove extension
         relative_path.set_extension("");
 
-        fs::write(wasm_binary_path, format!("{}", relative_path.display()))
-            .context("unable to write `.binpath`")?;
+        if !self.is_metawasm {
```
[the remark this answers is not in the document]
Comment (breathx) [reply utils/wasm-builder/src/wasm_project.rs]: Huge space for future refactoring

File: pallets/gear/rpc/src/lib.rs (hunk 3 above)
Comment (breathx) [reply pallets/gear/rpc/src/lib.rs:369]: Just for confidence that nothing will be changed in memory of program during metahash fn (we can't restrict its realisation)

File: core/src/code.rs (hunk 1 above)
Comment (breathx) [reply core/src/code.rs:252]: Seems `new_raw` doesn't have tests also, but both of them are only for non-production solution: tests and benches

**Pull request merged**